Added missing GTFS Extended Route Types

robertgregg3 opened the pull request:
- Added the route types to the gtfs.yml file 
- ~and then added logic to display the route_type field as an optgroup~
- Create a new selector for GTFS route types, with extended route types organized as a tree hierarchy (bus, metro, etc) that is not expanded by default to reduce screen clutter while searching for extended route types.

Fix #645

landonreed commented on lib/editor/components/EditorInput.js:
```diff
@@ -300,6 +300,8 @@ export default class EditorInput extends Component<Props> {
       case 'DROPDOWN':
+        const standard = 'Standard'
+        const extended = 'Extended'
         return (
```
To fix the flow type issues and simplify things, I would add something like the following:
```javascript
if (!field.options) {
  console.warn(`Field `field.name` does not have options configured, but is marked as inputType=DROPDOWN`)
  return null
}
```

robertgregg3 replied:
Where do you see the flow issues?  When I load the project (before typing any code, I get these kind of squigglies from the beginning of the render method:
![image](https://user-images.githubusercontent.com/4264701/114189718-68cba580-9942-11eb-8d3f-7f99253166f5.png)

robertgregg3 replied:
When I run `flow status` I get an error:
![image](https://user-images.githubusercontent.com/4264701/114190460-3d958600-9943-11eb-8393-94603ebb983f.png)

landonreed replied:
The flow errors are showing up in the GitHub builds: https://github.com/ibi-group/datatools-ui/pull/656/checks?check_run_id=2298915284

Locally, you should run `yarn flow`.

landonreed commented on gtfs.yml:
```diff
@@ -290,6 +290,156 @@
       text: Gondola
     - value: 7
       text: Funicular
```
Could you also add options 11 and 12 (they were recently added to the standard set)? See the spec for `routes.txt`: https://developers.google.com/transit/gtfs/reference#routestxt for more info

robertgregg3 replied:
Done!

landonreed commented on lib/editor/components/EditorInput.js:
```diff
@@ -314,7 +316,18 @@ export default class EditorInput extends Component<Props> {
+                  {field.options.slice(8).map(o => (
+                    <option value={o.value} key={o.value}>{o.text} ({o.value})</option>))}
+                </optgroup>}
+              )}
```
I would recommend placing this stuff (including lines 303 and 304) into a `_renderRouteTypeOptions` method to encapsulate it a bit more. Once that's done, I would recommend you use a ternary here instead of making the check for field name twice:
```javascript
field.name === 'route_type'
  ? this._renderRouteTypeOptions(options)
  : field.options.map(o => (<option value={o.value} key={o.value}>{o.text || o.value}</option>))
```

landonreed reviewed (changes requested):
A few changes, but this is looking good!

binh-dam-ibigroup commented on gtfs.yml:
```diff
@@ -290,6 +290,156 @@
       text: Gondola
     - value: 7
       text: Funicular
+    - value: 100
+      text: Railway Service
```
Why is this value duplicated?

robertgregg3 replied:
Good spot. And there was a word missing. Going to check the rest now.

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -314,7 +316,18 @@ export default class EditorInput extends Component<Props> {
-              {field.options && field.options.map(o => (<option value={o.value} key={o.value}>{o.text || o.value}</option>))}
+              {field.name === 'route_type' &&
+                <optgroup label={standard}>
+                  {field.options.slice(0, 7).map(o => (
```
Because `o` shows up so many times, I would write `.map(({ text, value }) => { ...`, and maybe also extract the `<optgroup>` rendering to its own method because of repetition.

robertgregg3 commented:
### I have added:
- a group option in the gtfs file
- added a _renderRouteTypeOptions method to dynamically render the opt groups depending on the group value (Standard or Extended)

landonreed reviewed (commented):
@robertgregg3, what prompted the addition of the `group` field in `gtfs.yml`? I'm not sure it's needed, given that we effectively know that types 0-12 are standard and those after are extended. I liked the way you had handled it previously.

Also, `Trolleybus (11)` and `Monorail (12)` should probably be in standard and moved up in the list. These are currently shown as extended (800 and 405), but I think that's probably just because the extended list has not been updated. We could keep them both, but that might be confusing.

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -118,6 +118,33 @@ export default class EditorInput extends Component<Props> {
+    const extendedTypes = []
+    const routeTypes = ['Standard', 'Extended']
+    const fieldOptions = field.options
+    for (const option in fieldOptions) {
+      if (fieldOptions[option].group === 'Standard') {
+        standardTypes.push(fieldOptions[option])
+      } else {
+        extendedTypes.push(fieldOptions[option])
+      }
+    }
```
This could be simplified:
```js
const standardTypes = Object.values(fieldOptions).filter(fieldOption => fieldOption.group === 'Standard')
// same for extendedTypes
```

binh-dam-ibigroup replied:
Even better, in the `return` statement, try this:
```js
return routeTypes
.map(type => Object.values(fieldOptions).filter(/*see above*/))
.map(routeTypes => {
  <optgroup>
    {routeTypes.map(routeType => /*what you already have for each std/extended types*/)}
  </optgroup>
})

robertgregg3 replied:
Nice, I like it! Thanks Binh

binh-dam-ibigroup reviewed (changes requested):
Almost there, see how the simplification suggestion would work with @landonreed's question regarding the groups of route types.

landonreed commented:
Commentary is a little muddy here, so I'm going to add a PR and share my recs there.

mvanlaar commented:
> @robertgregg3, what prompted the addition of the `group` field in `gtfs.yml`? I'm not sure it's needed, given that we effectively know that types 0-12 are standard and those after are extended. I liked the way you had handled it previously.
> 
> Also, `Trolleybus (11)` and `Monorail (12)` should probably be in standard and moved up in the list. These are currently shown as extended (800 and 405), but I think that's probably just because the extended list has not been updated. We could keep them both, but that might be confusing.

My opinion on this would be there will no code changes needed in the future if somethings changes to the spec. Only a edit to the gtfs.yml file. Also some companies can changes which items are for them standard and what are extended without changing code but only the gtfs.yml file.

landonreed commented:
@robertgregg3, take a look at my changes in https://github.com/ibi-group/datatools-ui/pull/659. Feel free to review and merge here if they look OK to you. And let me know if you have any questions!

robertgregg3 commented:
> @robertgregg3, what prompted the addition of the `group` field in `gtfs.yml`? I'm not sure it's needed, given that we effectively know that types 0-12 are standard and those after are extended. I liked the way you had handled it previously.
> 
@landonreed I had an interesting time trying to create an drop down with multiple optgroups (in React).  The form control in the DROPDOWN case appears to be looking for one nested div of options. When I tried to render 2 opt groups within one drop down I was getting the JSX error about it needing a surrounding div tag or React.Fragment. 

When researching I cam across others with similar issues in React. I found a solution that used grouping via the options data so that is why I added the "group" to the gtfs.yml route_type options.  

I did kind of prefer the other option and there was most likely a simple face palm solution.

robertgregg3 commented:
@landonreed Thanks for sharing your solution. Looks great

landonreed commented:
Re: @mvanlaar's point about avoiding a code change, yes, there may be some value in that. But these `route_type` values and their classification as Standard vs. Extended are intended to accurately represent the GTFS specification, rather than a third party's interpretation of what is "standard." I think we should stick with the slice method for now. Since both of these files (`gtfs.yml` and `EditorInput.js`) are committed to the repo (and there's no way to substitute alternative files for them), I don't see a major difference in the "no-code-change" (yml) vs. the js approach, so I'm inclined to stick with the current approach.

robertgregg3 commented:
@binh-dam-ibigroup the changes are from Landons Merged commit

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -1,6 +1,6 @@
 // @flow
 
-import React, {Component} from 'react'
+import * as React from 'react'
```
Can you use the previous import nomenclature? (unless others want to override me).

landonreed replied:
No, because referencing React.Node requires this import approach.

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -118,31 +118,26 @@ export default class EditorInput extends Component<Props> {
-        standardTypes.push(fieldOptions[option])
-      } else {
-        extendedTypes.push(fieldOptions[option])
+  _renderRouteTypeOptions = (field: GtfsSpecField): React.Node => {
```
I think the more accurate output type would be `Array<React.Node>` (non-blocking)

landonreed replied:
```suggestion
  _renderRouteTypeOptions = (field: GtfsSpecField): Array<React.Node> => {
```

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -118,6 +118,28 @@ export default class EditorInput extends Component<Props> {
+    const routeTypes = [
+      {
+        label: 'Standard',
+        options: options.slice(0, 9).map(toOption)
```
Using a filter on the range for `value` may be easier to read. As written, it looks like the array indexes and the comment above contradict each other.
```suggestion
        options: options.filter(opt => opt.value < 100).map(toOption)
```

binh-dam-ibigroup commented on lib/editor/components/EditorInput.js:
```diff
@@ -118,6 +118,28 @@ export default class EditorInput extends Component<Props> {
+      },
+      {
+        label: 'Extended',
+        options: options.slice(10).map(toOption)
```
```suggestion
        options: options.filter(opt => opt.value >= 100).map(toOption)
```

binh-dam-ibigroup reviewed (approved):
Conditional approval on using a filter for separating the standard vs. extended types for improved readability.

binh-dam-ibigroup commented:
@robertgregg3 Please fix the failing build...

binh-dam-ibigroup commented:
This PR seemed to be blocked pending #659, but those changes have been merged since. Removing the "Blocked" label.

binh-dam-ibigroup commented:
We can probably start reviewing this PR at this point. There may be some more refactorings possible that would touch parts that are not relevant to this PR, but let me know if you want to make them.

miles-grant-ibigroup reviewed (approved):
This actually looks quite good! The dropdown feels a bit janky but given that we're using a library this is perhaps to be expected. Other than a few style changes I think this might actually be good to go!

Thanks for updating the e2e tests as well.

binh-dam-ibigroup commented on lib/style.css:
```diff
@@ -879,3 +879,111 @@ h4.line:after {
+
+/* Remove border and background of tags
+ * (a "tag" is a selected item displayed inside the input portion of the selector). */
+.route-type-select .dropdown-trigger .tag {
```
Which bold are you referring to?

binh-dam-ibigroup replied:
Oh I see, in FF.

binh-dam-ibigroup replied:
Thanks, addressed in https://github.com/ibi-group/datatools-ui/pull/656/commits/bc14fedc2b8390f387d73beb60777a528e0bf857.

philip-cline reviewed (approved):
Looks good to me!
